src/GraphQL.MemoryCache/AutomaticPersistedQueries/AutomaticPersistedQueriesExecution.cs

```diff
@@ -0,0 +1,120 @@
+    /// </summary>
+    protected virtual (string? Hash, string? Version, bool Enabled) GetAPQProperties(Inputs? extensions)
+    {
+        string? hashResult = null;
+        string? versionResult = null;
+        bool enabled = false;
+
+        if (extensions?.TryGetValue("persistedQuery", out var persistedQueryObject) ?? false)
+        {
+            enabled = true;
+
+            if (persistedQueryObject is Dictionary<string, object> persistedQuery)
+            {
+                if (persistedQuery.TryGetValue("sha256Hash", out var sha256HashObject) && sha256HashObject is string sha256Hash && !string.IsNullOrWhiteSpace(sha256Hash))
+                {
+                    hashResult = sha256Hash;
+                }
+
+                if (persistedQuery.TryGetValue("version", out var versionObject) && versionObject is string version && !string.IsNullOrWhiteSpace(version))
+                {
+                    versionResult = version;
+                }
+            }
+        }
+
+        return (hashResult, versionResult, enabled);
+    }
+
+    /// <summary>
+    /// Compute SHA256 hash.
+    /// </summary>
+    protected virtual string ComputeSHA256(string input)
+    {
+        var bytes = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(input));
+
+        var builder = new StringBuilder();
+        foreach (var item in bytes)
+        {
+            builder.Append(item.ToString("x2"));
+        }
+
+        return builder.ToString();
+    }
+
+    /// <summary>
+    /// Create <see cref="ExecutionResult"/> with provided error.
+    /// </summary>
+    protected virtual ExecutionResult CreateExecutionResult(ExecutionError error) => new ExecutionResult { Errors = new ExecutionErrors { error } };
+
+    /// <inheritdoc/>
+    public async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        var apq = GetAPQProperties(options.Extensions);
+
+        if (apq.Enabled && apq.Version != SUPPORTED_VERSION)
+        {
+            return CreateExecutionResult(new PersistedQueryUnsupportedVersionError(apq.Version));
+        }
+
+        if (string.IsNullOrWhiteSpace(options.Query))
+        {
+            if (string.IsNullOrWhiteSpace(apq.Hash))
+            {
+                return CreateExecutionResult(new ExecutionError("GraphQL query is missing."));
+            }
+            else
+            {
+                var queryFromCache = await _cache.GetAsync(apq.Hash!).ConfigureAwait(false);
+
+                if (queryFromCache == null)
+                {
+                    return CreateExecutionResult(new PersistedQueryNotFoundError(apq.Hash!));
+                }
+                else
+                {
+                    options.Query = queryFromCache;
+                }
+            }
+        }
+        else if (!string.IsNullOrWhiteSpace(apq.Hash))
+        {
+            if (apq.Hash!.Equals(ComputeSHA256(options.Query!), StringComparison.InvariantCultureIgnoreCase))
```

Comment:
I suggest to switch from comparing characters to comparing bytes.

Reply:
`StringComparison.Ordinal` ?

Reply:
No, comparing raw bytes i.e remove byte[] -> string conversion from `ComputeSHA256` but convert apq.Hash to `byte[]`. It can be done with ArrayPool.

Reply:
With the current inheritance pattern, which we already discussed, I would leave as-is.